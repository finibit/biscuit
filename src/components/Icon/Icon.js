import React from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import styles, { utils } from '../../themes'

const upArrowStyles = css`
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-bottom: 4px solid ${(props) => utils.resolveColor(props.theme, props.$element, props.$color)};
`

const downArrowStyles = css`
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-top: 4px solid ${(props) => utils.resolveColor(props.theme, props.$element, props.$color)};
`

const rightArrowStyles = css`
  width: 0;
  height: 0;
  border-top: 4px solid transparent;
  border-bottom: 4px solid transparent;
  border-left: 4px solid ${(props) => utils.resolveColor(props.theme, props.$element, props.$color)};
`

const leftArrowStyles = css`
  width: 0;
  height: 0;
  border-top: 4px solid transparent;
  border-bottom: 4px solid transparent;
  border-right: 4px solid ${(props) => utils.resolveColor(props.theme, props.$element, props.$color)};
`

const checkMarkStyles = css`
  &:after {
    content: '';
    display: block;
    width: 4px;
    height: 9px;
    border: solid ${(props) => utils.resolveColor(props.theme, props.$element, props.$color)};
    border-width: 0 2px 2px 0;
    transform: rotate(45deg);
  }
`

const IconStyled = styled.div`
  ${(props) => {
    switch (props.$type) {
    case 'arrow-up':
      return upArrowStyles
    case 'arrow-left':
      return leftArrowStyles
    case 'arrow-down':
      return downArrowStyles
    case 'arrow-right':
      return rightArrowStyles
    case 'check-mark':
      return checkMarkStyles
    default:
      return ''
    }
  }}

  display: inline-block;

  ${styles.spacing}
  ${(props) => props.$css}
`

/** Semantic CSS graphics. */
const Icon = (props) => {
  const {
    type,
    color,
    margin,
    padding,
    themeElement,
    css,
    ...rest
  } = props

  return (
    <IconStyled
      $element={themeElement}
      $type={type}
      $color={color}
      $margin={margin}
      $padding={padding}
      $css={css}
      {...rest}
    />
  )
}

Icon.propTypes = {
  /** Type of the icon. */
  type: PropTypes.oneOf([
    'arrow-up',
    'arrow-left',
    'arrow-down',
    'arrow-right',
    'check-mark'
  ]).isRequired,

  /** Color of the icon. */
  color: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      shade: PropTypes.number,
      tint: PropTypes.number
    })
  ]),

  /** The amount of margin around the icon. */
  margin: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),

  /** The amount of padding around the icon. */
  padding: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.object,
    PropTypes.array
  ]),

  /** Theme element. */
  themeElement: PropTypes.string,

  /** Custom styles passed to styled-components. */
  css: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
}

Icon.defaultProps = {
  color: 0,
  margin: 'none',
  padding: 'none',
  themeElement: 'Icon',
  css: null
}

export default Icon
