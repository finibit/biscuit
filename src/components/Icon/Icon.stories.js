import React from 'react'
import { storiesOf } from '@storybook/react'

import { Theme, Icon } from '../../../src'

storiesOf('Icon', module)
	.add('types', () => (
		<Theme>
			<Icon type="arrow-up" margin="xs" />
			<Icon type="arrow-left" margin="xs" />
			<Icon type="arrow-down" margin="xs" />
			<Icon type="arrow-right" margin="xs" />
			<Icon type="check-mark" margin="xs" />
		</Theme>
	))
