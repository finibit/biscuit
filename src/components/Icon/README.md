### Usage

#### Type

```jsx
<Icon type="check-mark" />
```
