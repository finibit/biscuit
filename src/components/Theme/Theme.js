import React from 'react'
import PropTypes from 'prop-types'
import { ThemeProvider, createGlobalStyle } from 'styled-components'
import baseTheme from '../../themes/baseTheme'
import mergeTheme from '../../themes/themeUtils'

const GlobalStyle = createGlobalStyle`
	html {
		box-sizing: border-box;
	}

	html, body {
		width: 100%;
		height: 100%;
		margin: 0;
		padding: 0;
		font-size: ${(props) => props.theme.global.baseSize}px;
		font-family: ${(props) => props.theme.global.fontFamily};
	}

	*, *:before, *:after {
		box-sizing: inherit;
	}
`

const Theme = (props) => (
	<ThemeProvider theme={mergeTheme(baseTheme, props.theme)}>
		<div>
			<GlobalStyle />
			{props.children}
		</div>
	</ThemeProvider>
)

Theme.propTypes = {
	children: PropTypes.node,
	theme: PropTypes.oneOfType([PropTypes.object]),
}

Theme.defaultProps = {
	children: null,
	theme: baseTheme,
}

export default Theme
