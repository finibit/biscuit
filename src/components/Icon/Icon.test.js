import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Theme from '../Theme'
import Icon from '../Icon'

describe('Icon', () => {
  test('is rendered', () => {
    const component = renderer.create((
      <Theme>
        <Icon type='arrow-up' />
        <Icon type='arrow-left' />
        <Icon type='arrow-down' />
        <Icon type='arrow-right' />
        <Icon type='check-mark' />
      </Theme>
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
