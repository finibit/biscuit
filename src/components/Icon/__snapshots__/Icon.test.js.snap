// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon is rendered 1`] = `
.c0 {
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-bottom: 4px solid #0a0a0a;
  display: inline-block;
  margin: 0;
  padding: 0;
}

.c1 {
  width: 0;
  height: 0;
  border-top: 4px solid transparent;
  border-bottom: 4px solid transparent;
  border-right: 4px solid #0a0a0a;
  display: inline-block;
  margin: 0;
  padding: 0;
}

.c2 {
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-top: 4px solid #0a0a0a;
  display: inline-block;
  margin: 0;
  padding: 0;
}

.c3 {
  width: 0;
  height: 0;
  border-top: 4px solid transparent;
  border-bottom: 4px solid transparent;
  border-left: 4px solid #0a0a0a;
  display: inline-block;
  margin: 0;
  padding: 0;
}

.c4 {
  display: inline-block;
  margin: 0;
  padding: 0;
}

.c4:after {
  content: '';
  display: block;
  width: 4px;
  height: 9px;
  border: solid #0a0a0a;
  border-width: 0 2px 2px 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

@media only screen and (min-width:48rem) {
  .c0 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:64rem) {
  .c0 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:75rem) {
  .c0 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:48rem) {
  .c1 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:64rem) {
  .c1 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:75rem) {
  .c1 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:48rem) {
  .c2 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:64rem) {
  .c2 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:75rem) {
  .c2 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:48rem) {
  .c3 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:64rem) {
  .c3 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:75rem) {
  .c3 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:48rem) {
  .c4 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:64rem) {
  .c4 {
    margin: 0;
    padding: 0;
  }
}

@media only screen and (min-width:75rem) {
  .c4 {
    margin: 0;
    padding: 0;
  }
}

<div>
  <div
    className="c0"
  />
  <div
    className="c1"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
</div>
`;
